### PR TITLE
Add TokenBridge deployment

### PR DESCRIPTION
**Pull request by nambrot**
### Description

This PR adds deployment tooling and artifacts for the Token Bridge

### Drive-by changes

Move tooling under a `middleware` folder (and not `interchain`)

### Backward compatibility

Yes

### Testing

_What kind of testing have these changes undergone?_

Manual

### CODE REVIEW

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  Mock = 'Mock',
+  Circle = 'Circle',
+}
+
+export interface CircleBridgeAdapterConfig {
+  type: BridgeAdapterType.Circle;
+  circleBridgeAddress: string;
+  messageTransmitterAddress: string;
+  usdcAddress: string;
+  circleDomainMapping: {
```
Comment (yorhodes) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:32]: is this a mapping?

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    hyperlaneDomain: number;
+    circleDomain: number;
+  }[];
+}
+
+interface MockTokenBridgeAdapterConfig {
+  type: BridgeAdapterType.Mock;
+  mockTokenAddress: string;
```
Comment (yorhodes) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:40]: can't you just instantiate a mock of the real type with a fake usdc address?

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+export type BridgeAdapterConfig =
+  | CircleBridgeAdapterConfig
+  | MockTokenBridgeAdapterConfig;
+
+export type TokenBridgeConfig = RouterConfig & {
+  bridgeAdapterConfigs: BridgeAdapterConfig[];
```
Comment (yorhodes) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:48]: don't think this should include "mock types"

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  }
+
+  async enrollRemoteRouters(
+    contractsMap: ChainMap<Chain, TokenBridgeContracts>,
+  ): Promise<void> {
+    // Enroll the TokenBridgeRouter with each other
+    await super.enrollRemoteRouters(contractsMap);
+
+    // Enroll the circle adapters with each other
+    const deployedChains = Object.keys(contractsMap);
```
Comment (yorhodes) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:75]: isn't this basically just `RouterDeployer.enrollRemoteRouters` for the `CircleBridgeAdapter`?

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                adapterConfig.messageTransmitterAddress,
+                router.address,
+              ],
+            );
+          const circleBridgeAdapter = await this.deployContract(
+            chain,
+            'circleBridgeAdapter',
+            [],
+            {
+              create2Salt: this.create2salt + 'CircleBridgeAdapter',
```
Comment (yorhodes) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:157]: why salt these differently?

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    config: TokenBridgeConfig,
+  ): Promise<TokenBridgeContracts> {
+    const cc = this.multiProvider.getChainConnection(chain);
+    const initCalldata =
+      TokenBridgeRouter__factory.createInterface().encodeFunctionData(
+        'initialize',
+        [config.owner, config.connectionManager, config.interchainGasPaymaster],
+      );
+    const router = await this.deployContract(chain, 'router', [], {
+      create2Salt: this.create2salt + 'TokenBridgeRouter',
```
Comment (yorhodes) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:133]: why use a different salt?

File: typescript/sdk/src/middleware/tokenbridge.hardhat-test.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    multiProvider = getTestMultiProvider(signer);
+
+    const coreDeployer = new TestCoreDeployer(multiProvider);
+    const coreContractsMaps = await coreDeployer.deploy();
+    coreApp = new TestCoreApp(coreContractsMaps, multiProvider);
+
+    const mockTokenF = new MockToken__factory(signer);
+    mockToken = await mockTokenF.deploy();
+
+    config = coreApp.extendWithConnectionClientConfig(
```
Comment (yorhodes) [inline typescript/sdk/src/middleware/tokenbridge.hardhat-test.ts:54]: these are getting a bit illegible

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts (hunk 1 above)
Comment (nambrot) [reply typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:32]: Is it not?

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts (hunk 4 above)
Comment (nambrot) [reply typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:75]: Yes it is, but I didn't want to extract the logic out to be independent of the key access. But now that you said it, i guess i can just pass a modified "contractsMap"

File: typescript/infra/scripts/middleware/deploy-token-bridge.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    multiProvider,
+    objMap(circleBridgeAdapterConfig, (chain, conf) => ({
+      bridgeAdapterConfigs: [conf],
+      ...ownerConfigMap[chain],
+    })),
+    core,
+    'asdssds',
+  );
+
+  await deployWithArtifacts(dir, tokenBridgeFactories, deployer);
```
Comment (jmrossy) [inline typescript/infra/scripts/middleware/deploy-token-bridge.ts:43]: Should we add these addresses to the SDK now like we did for other middleware?

File: typescript/infra/scripts/middleware/deploy-token-bridge.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  const ownerConfigMap = await getConfiguration(environment, multiProvider);
+
+  const deployer = new TokenBridgeDeployer(
+    multiProvider,
+    objMap(circleBridgeAdapterConfig, (chain, conf) => ({
+      bridgeAdapterConfigs: [conf],
+      ...ownerConfigMap[chain],
+    })),
+    core,
+    'asdssds',
```
Comment (jmrossy) [inline typescript/infra/scripts/middleware/deploy-token-bridge.ts:40]: Good chance for a better salt :)

File: typescript/infra/src/testrecipient/index.ts [hunk 10]
```diff
@@ -30,8 +37,15 @@ export class TestRecipientDeployer<
       [],
       { create2Salt: 'testtest32' },
     );
+    const TestTokenRecipient = await this.deployContract(
+      chain,
+      'TestTokenRecipient',
+      [],
+      { create2Salt: 'testtest32' },
```
Comment (jmrossy) [inline typescript/infra/src/testrecipient/index.ts:44]: ditto, where are these salts coming from haha?

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+  Chain,
+  TokenBridgeConfig,
+  TokenBridgeContracts,
+  TokenBridgeFactories
+> {
+  constructor(
+    multiProvider: MultiProvider<Chain>,
+    configMap: ChainMap<Chain, TokenBridgeConfig>,
+    protected core: HyperlaneCore<Chain>,
+    protected create2salt = 'tokenbridgedeployersalt',
```
Comment (jmrossy) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:63]: camel case?

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+          ) {
+            continue;
+          }
+          const current = await contracts.circleBridgeAdapter.routers(
+            remoteDomain,
+          );
+          const expected = utils.addressToBytes32(
+            contractsMap[remote].circleBridgeAdapter!.address,
+          );
+          if (current !== expected) {
```
Comment (jmrossy) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:97]: Code's going far to the right here, consider inverting this if (`if === then continue`)

File: typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    const router = await this.deployContract(chain, 'router', [], {
+      create2Salt: this.create2salt + 'TokenBridgeRouter',
+      initCalldata,
+    });
+
+    const bridgeAdapters: Partial<TokenBridgeContracts> = {};
+
+    for (const adapterConfig of config.bridgeAdapterConfigs) {
+      switch (adapterConfig.type) {
+        case BridgeAdapterType.Circle:
```
Comment (jmrossy) [inline typescript/sdk/src/deploy/middleware/TokenBridgeRouterDeployer.ts:141]: Nit, switches in JS are actually more verbose and create more indentation than just a series of if statements :)

File: typescript/sdk/src/tokenBridge.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import {
```
Comment (jmrossy) [inline typescript/sdk/src/tokenBridge.ts:1]: Consider moving this to `/middleware` ? Or making a tokenBridge dir and keeping everything together?

File: typescript/sdk/src/middleware/tokenbridge.hardhat-test.ts (hunk 7 above)
Comment (nambrot) [reply typescript/sdk/src/middleware/tokenbridge.hardhat-test.ts:54]: idk if there is an easier way though?

File: typescript/infra/scripts/middleware/deploy-token-bridge.ts (hunk 8 above)
Comment (nambrot) [reply typescript/infra/scripts/middleware/deploy-token-bridge.ts:43]: I feel like in an ideal world we would just write them directly, but for now, I just do the `mergeWithSdkContractAddressArtifacts` function right?

Comment (jmrossy) [reply typescript/infra/scripts/middleware/deploy-token-bridge.ts:43]: Yes and you'll also want to expose it in the `hyperlaneCoreAddresses` map:
https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1210/files#diff-8a8d0a4753bc69e9ad052d746b5d3704618d541a46034d2e0449bdb5f651958dR20

Comment (nambrot) [reply typescript/infra/scripts/middleware/deploy-token-bridge.ts:44]: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1215/commits/a08c14ffa7034077b56d7c497fbeadf049d28ab0

File: typescript/sdk/src/middleware/tokenbridge.hardhat-test.ts (hunk 7 above)
Comment (jmrossy) [reply typescript/sdk/src/middleware/tokenbridge.hardhat-test.ts:68]: If/when we overhaul the SDK, one idea would be a set of functions with consistent naming that extend configs with different things. e.g. `withClientConfig`, `withOwnerAddress`, etc.

**Review by jmrossy — APPROVED** [review]
Did not review solidity changes but js changes lgtm

File: typescript/infra/scripts/middleware/deploy-token-bridge.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    objMap(circleBridgeAdapterConfig, (chain, conf) => ({
+      bridgeAdapterConfigs: [conf],
+      ...ownerConfigMap[chain],
+    })),
+    core,
+    'TokenBridgeDeploy2',
+  );
+
+  await deployWithArtifacts(dir, tokenBridgeFactories, deployer);
+  await mergeWithSdkContractAddressArtifacts(environment);
```
[the remark this answers is not in the document]
Comment (nambrot) [reply typescript/infra/scripts/middleware/deploy-token-bridge.ts]: np

**Pull request merged**